testenv/saml.py

```diff
@@ -395,3 +401,239 @@ def create_error_response(data, response_status):
     )
     response.append(issuer)
     return response
+
+
+# Metadata
+
+class EntityDescriptor(SamlMixin):
+    saml_type = 'md'
+
+
+class IDPSSODescriptor(SamlMixin):
+    saml_type = 'md'
+    defaults = {
+        'protocolSupportEnumeration': SAML,
+    }
+
+
+class SPSSODescriptor(SamlMixin):
+    saml_type = 'md'
+    defaults = {
+        'protocolSupportEnumeration': SAML,
+    }
+
+
+class NameIDFormat(SamlMixin):
+    saml_type = 'md'
+
+
+class KeyDescriptor(SamlMixin):
+    saml_type = 'md'
+
+
+class KeyInfo(SamlMixin):
+    saml_type = 'ds'
+
+
+class X509Data(SamlMixin):
+    saml_type = 'ds'
+
+
+class X509Certificate(SamlMixin):
+    saml_type = 'ds'
+
+
+class SingleSignOnService(SamlMixin):
+    saml_type = 'md'
+
+
+class SingleLogoutService(SamlMixin):
+    saml_type = 'md'
+
+
+class Organization(SamlMixin):
+    saml_type = 'md'
+
+
+class OrganizationName(SamlMixin):
+    saml_type = 'md'
+
+
+class OrganizationURL(SamlMixin):
+    saml_type = 'md'
+
+
+class AssertionConsumerService(SamlMixin):
+    saml_type = 'md'
+    defaults = {
+        'Binding': BINDING_HTTP_POST
+    }
+
+
+class AttributeConsumingService(SamlMixin):
+    saml_type = 'md'
+
+
+class ServiceName(SamlMixin):
+    saml_type = 'md'
+
+
+class RequestedAttribute(SamlMixin):
+    saml_type = 'md'
+
+
+
+def create_idp_metadata(
+    entity_id,
+    want_authn_requests_signed,
+    keys=[],
+    single_sign_on_services=[],
+    single_logout_services=[],
+    attributes=[],
```

Comment:
Maybe we should replace these mutable default arguments with `None`, and check their value inside the function, to avoid possible data persistency across function invocations.